### PR TITLE
Introduce basic shell scripts

### PR DESCRIPTION
**Pull request by Yonas-net**

### CODE REVIEW

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
I don’t insist on re-writing in Go, but IMAO it would make the following easier.

* Parametrize host, port, user and password of both Icinga 2 and IW2 and `/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/conf.d`
* Randomize names of created objects
* Make all output human-readable

File: script.sh [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  rm $confdPath/$object".conf"
+  curl -k -s -u $icinga2Auth -H 'Accept: application/json' -X POST "https://$icinga2Host/v1/actions/restart-process"
+  echo
+
+  # Querying Icinga Web 2 for host created using config file
+  sleep 10;
+  curl=("curl" -skSu "$icingaweb2Auth" -H 'Accept: application/json' -X GET "http://$icingaweb2Host/icingaweb2/monitoring/list/hosts?host=$object&modifyFilter=1")
+  opt=$("${curl[@]}")
+  if [[ -n $opt ]]; then
+    printf "\nHost %s object can be found in Icinga Web 2.\n" $object
```
Comment (Al2Klimov) [inline script.sh:58]: Those look fine, but...

File: script.sh [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      zone = "ha-master1"
+    }
+EOF
+
+  curl -skSu $icinga2Auth -H 'Accept: application/json' -X POST "https://$icinga2Host/v1/actions/restart-process"
+  echo
+
+  sleep 60
+  rm $confdPath/$object".conf"
+  curl -k -s -u $icinga2Auth -H 'Accept: application/json' -X POST "https://$icinga2Host/v1/actions/restart-process"
```
Comment (Al2Klimov) [inline script.sh:50]: ... the script shall also echo what **these ones** are doing.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: script.sh (hunk 2 above)
Comment (Al2Klimov) [reply script.sh]: I also meant sleep and rm.

File: script.sh [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  printf "\n\t--i2Host Icinga 2 Host localhost:5665\n"
+  printf "\t--i2Auth Icinga 2 API authentication root:icinga\n"
+  printf "\t--iw2Host Icinga Web 2 url\n"
+  printf "\t--iw2Auth Icing Web 2 authentication username:password\n"
+  printf "\t--cfdPath Absolute path to icinga2 conf.d directory\n"
+  exit 1
+}
+
+while echo $1 | grep -q ^-; do
+  declare "$(echo $1 | sed 's/^-//')=$2"
```
Comment (Al2Klimov) [inline script.sh:15]: You’re already setting variables here...

File: script.sh [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  shift
+  shift
+done
+
+icinga2Host="$i2Host"
+icinga2Auth="$i2Auth"
+icingaweb2Host="$iw2Host"
+icingaweb2Auth="$iw2Auth"
+confdPath="$cfdPath"
```
Comment (Al2Klimov) [inline script.sh:24]: ... so why are you setting additional ones with the same stuff?

File: script.sh [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  curl=("curl" -skSu "$icinga2Auth" -H 'Accept: application/json' -X POST "https://$icinga2Host/v1/actions/restart-process")
+  opt=$("${curl[@]}")
+  if [[ -n $opt ]]; then
+    printf "\nIcinga 2 daemon successfully restarted.\n"
+  else
+    printf "\nFailed to restart Icinga 2 daemon.\n"
+  fi
+
+  # Querying Icinga Web 2 for host created using config file
+  sleep 10
```
Comment (Al2Klimov) [inline script.sh:68]: Please log all sleeps...

File: script.sh [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+  cat > $confdPath/$object".conf" <<-"EOF"
+    object Host "host-objects" {
+      display_name = "icingaservice"
+      check_command = "dummy"
+      enable_active_checks = false
+      enable_notifications = "1.000000"
+      zone = "ha-master1"
+    }
+EOF
```
Comment (Al2Klimov) [inline script.sh:44]: ... and this one.

File: script.sh [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  if [[ -n $opt ]]; then
+    printf "\nHost %s object can be found in Icinga Web 2.\n" $object
+  else
+    printf "\nHost %s object doesn't exists in Icinga Web 2.\n" $object
+  fi
+
+  # Create host via the API
+  curl -skSu $icinga2Auth -H 'Accept: application/json' -X PUT "https://$icinga2Host/v1/objects/hosts/$object" \
+    -d '{ "attrs": { "check_command": "hostalive", "address": "10.211.55.12", "display_name": "icingaservice", "zone": "ha-master1", "vars": { "os": "Linux"} } }'
+  echo
```
Comment (Al2Klimov) [inline script.sh:80]: ... and all curls.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Have a look what you're doing more than once and put that in functions.

File: script.sh [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+fi
+
+CreateHost() {
+  curl=("curl" -skSu "$i2Auth" -H 'Accept: application/json' -X PUT "$1")
+  opt=$("${curl[@]}")
+  if [[ -n $opt ]]; then
+    return 0
+  else
+    return 1
+  fi
```
Comment (Al2Klimov) [inline script.sh:31]: Put the logs in the functions.

File: script.sh [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    object Host "host-objects" {
+      display_name = "icingaservice"
+      check_command = "dummy"
+      enable_active_checks = false
+      enable_notifications = "1.000000"
+      zone = "ha-master1"
+    }
+EOF
+
+  if ModifyObject "https://$i2Host/v1/actions/restart-process"; then
```
Comment (Al2Klimov) [inline script.sh:79]: Change your functions so they take only variable stuff, i.e. just the host name in this case.

File: script.sh [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+  shift 2
+done
+
+if [ -z "$i2Host" ] || [ -z "$i2Auth" ] || [ -z "$iw2Host" ] || [ -z "$iw2Auth" ] || [ -z "$cfdPath" ]; then
+  printf "\nSome of the required parameters are empty!\n"
+  PrintHelp
+fi
+
+CreateHost() {
+  curl=("curl" -skSu "$i2Auth" -H 'Accept: application/json' -X PUT "$1")
```
Comment (Al2Klimov) [inline script.sh:25]: This function takes `$1` – an URL...

File: script.sh [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+  else
+    printf "\nHost %s object doesn't exists in Icinga Web 2.\n" $object
+  fi
+
+  # Create host via the API
+  if CreateHost "https://$i2Host/v1/objects/hosts/$object" -d '{ "attrs": {
+    "check_command": "hostalive",
+    "address": "10.211.55.12",
+    "display_name": "icingaservice",
+    "vars": { "os": "Linux"} } }'; then
```
Comment (Al2Klimov) [inline script.sh:111]: ... which is `https://$i2Host/v1/objects/hosts/$object` in this case. What are `-d` `{...}` for in this case?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: script.sh (hunk 9 above)
Comment (Yonas-net) [reply script.sh:79]: How is this supposed to work? Restarting the process, scheduling a downtime and sending a notification is not the same even if all the queries use the same host.

Comment (Al2Klimov) [reply script.sh:79]: I mean:

1. Change this function, so you call it: `ModifyObject "$i2Host"`
2. Have a look at the others and change them the same way

Comment (Yonas-net) [reply script.sh:79]: The function is not only used to restart the process but also for something else, i.e. they do not have the same url e.g ``/v1/actions/schedule-downtime``.

Comment (Al2Klimov) [reply script.sh:79]: Now I see! o.o Name your functions after what they do.

File: script.sh (hunk 11 above)
Comment (Yonas-net) [reply script.sh]: For the associated host attributes, how else am I going to create a two hosts with different ``attrs``? I have just used incorrect parameter passing.

Comment (Al2Klimov) [reply script.sh]: Exactly. Correct your parameter passing.

**Review by Al2Klimov — COMMENTED** [review]
Run this and provide the complete output.

**Comment by Yonas-net** [comment]
```bash
./script.sh -i2Host "localhost:5665" -i2Auth "root:icinga" -iw2Host "10.211.55.14:80" -iw2Auth "icingaadmin:icinga" -cfdPath "/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/conf.d"
```

Output:

```bash
Creating Icinga 2 '/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/conf.d/host-23008.conf' config file.

Icinga 2 daemon successfully restarted.
Sleep until Icinga 2 has been successfully restarted.
Removing Icinga 2 '/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/conf.d/host-23008.conf' config file.

Icinga 2 daemon successfully restarted.
Sleep until Icinga 2 has been successfully restarted.

Host host-23008 object can be found in Icinga Web 2.

Host 'host-23008' object successfully created via API.
Sleep until Icinga 2 reloaded all necessary jobs.

Host host-23008 object can be found in Icinga Web 2.

Host host-23008 object can be found in Icinga 2.

Host 'icinga2' object successfully created via API.
Sleep until Host 'icinga2' object is visible in Icinga Web 2.

Icinga Web 2 host host-23008 object can be found.

Icinga 2 API host host-23008 object found.

Downtime of type 'Host' successfully scheduled.

Sending 'custom' Notification of type 'Host'.

Sleep until all created hosts are visible in Icinga Web 2.

Host 'host-23008' object successfully deleted.
Host 'icinga2' object successfully deleted.

Sleep until all deleted Hosts are also deleted from IDO.
```

**Comment by Al2Klimov** [comment]
Did you [just](https://github.com/Yonas-net/icinga-shell/pull/1#issuecomment-773962143) reproduce the issue this script is for?

**Comment by Yonas-net** [comment]
**v2.11.7**:

```bash
Creating Icinga 2 '/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/conf.d/host-30592.conf' config file.

Icinga 2 daemon successfully restarted.
Sleep until Icinga 2 has been successfully restarted.
Removing Icinga 2 '/Users/yhabteab/Workspace/icinga2/prefix/etc/icinga2/conf.d/host-30592.conf' config file.

Icinga 2 daemon successfully restarted.
Sleep until Icinga 2 has been successfully restarted.

Host host-30592 object doesn't exists in Icinga Web 2.

Host 'host-30592' object successfully created via API.
Sleep until Icinga 2 reloaded all necessary jobs.

Host host-30592 object doesn't exists in Icinga Web 2.

Host host-30592 object can be found in Icinga 2.

Host 'icinga2' object successfully created via API.
Sleep until Host 'icinga2' object is visible in Icinga Web 2.

Icinga Web 2 host 'icinga2' object can be found.

Icinga 2 API host 'icinga2' object found.

Downtime of type 'Host' successfully scheduled.

Sending 'custom' Notification of type 'Host'.

Sleep until all created hosts are visible in Icinga Web 2.

Host 'host-30592' object successfully deleted.
Host 'icinga2' object successfully deleted.

Sleep until all deleted Hosts are also deleted from IDO.
```

**Master**:

```bash
Coming soon 
```

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Squash.

**Review by Al2Klimov — APPROVED** [review]
I don’t like misusing bash for _actual programming_, but it seems to work.

**Pull request merged**